src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/FeatureSwitchManager.cs

```diff
@@ -726,6 +731,94 @@ private string GetResourceStringForAccessor(EcmaMethod method)
             return null;
         }
 
+        private static bool TryExpandTypeEquality(MethodIL methodIL, byte[] body, OpcodeFlags[] flags, int offset, string op, out int constant)
+        {
+            // We expect to see a sequence:
+            // ldtoken Foo
+            // call GetTypeFromHandle
+            // ldtoken Bar
+            // call GetTypeFromHandle
+            // -> offset points here
+            constant = 0;
+            const int SequenceLength = 20;
+            if (offset < SequenceLength)
+                return false;
+
+            if ((flags[offset - SequenceLength] & OpcodeFlags.InstructionStart) == 0)
+                return false;
+
+            ILReader reader = new ILReader(body, offset - SequenceLength);
+
+            TypeDesc type1 = ReadLdToken(ref reader, methodIL, flags);
+            if (type1 == null)
+                return false;
+
+            if (!ReadGetTypeFromHandle(ref reader, methodIL, flags))
+                return false;
+
+            TypeDesc type2 = ReadLdToken(ref reader, methodIL, flags);
+            if (type1 == null)
+                return false;
+
+            if (!ReadGetTypeFromHandle(ref reader, methodIL, flags))
+                return false;
+
+            // Dataflow runs on top of uninstantiated IL and we can't answer some questions there.
+            // Unfortunately this means dataflow will still see code that the rest of the system
+            // might have optimized away. It should not be a problem in practice.
+            if (type1.ContainsSignatureVariables() || type2.ContainsSignatureVariables())
+                return false;
+
+            // If either type is in canonical form, we cannot directly compare, but we could still be able
```

Comment:
Can we use `compareTypesForEquality` method from JIT/EE interface here? Or at least use the exact logic with same comments?

Reply:
Good point, that code also handles more cases. Moved to a shared place.